### PR TITLE
Better verification of remset

### PR DESCRIPTION
**Pull request by kdnilsen**
We do not need a remembered set at the start of a Global GC because Global GC scans all of old and all of young.  Because it is not necessary, the implementation of System.gc() may jump directly into marking without finishing the evacuation or coalescing and filling of previously identified mixed-evacuation candidates.  This means the remembered set may be in an incoherent state at the start of Global GC.  If this occurs, the verifier should not attempt to verify the remembered set at the start of Global GC.

---------
### Progress
- [ ] Change must not contain extraneous whitespace
- [ ] Commit message must refer to an issue
- [ ] Change must be properly reviewed (1 review required, with at least 1 [Committer](https://openjdk.org/bylaws#committer))

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/shenandoah.git pull/423/head:pull/423` \
`$ git checkout pull/423`

Update a local copy of the PR: \
`$ git checkout pull/423` \
`$ git pull https://git.openjdk.org/shenandoah.git pull/423/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 423`

View PR using the GUI difftool: \
`$ git pr show -t 423`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/shenandoah/pull/423.diff">https://git.openjdk.org/shenandoah/pull/423.diff</a>

</details>

### CODE REVIEW

**Comment by kdnilsen** [comment]
I'll leave this in draft mode until I have finished more extensive testing.

File: src/hotspot/share/gc/shenandoah/shenandoahVerifier.cpp [hunk 1]
```diff
@@ -1354,7 +1382,18 @@ void ShenandoahVerifier::verify_rem_set_before_mark() {
 
   log_debug(gc)("Verifying remembered set at %s mark", _heap->doing_mixed_evacuations()? "mixed": "young");
```
Comment (earthling-amzn) [inline src/hotspot/share/gc/shenandoah/shenandoahVerifier.cpp:1383]: Might change this log message to something like:
```C++
log_debug(gc)("Verifying remembered set at %s%s mark", _heap->active_generation()->name(), _heap->doing_mixed_evacuations()? " (mixed)": "");
```

File: src/hotspot/share/gc/shenandoah/shenandoahVerifier.cpp [hunk 2]
```diff
@@ -1333,12 +1333,40 @@ void ShenandoahVerifier::help_verify_region_rem_set(ShenandoahHeapRegion* r, She
...
+        HeapWord* orig_obj_addr = obj_addr;
         obj_addr = ctx->get_next_marked_addr(obj_addr, tams);
+        confirm_filled(orig_obj_addr, obj_addr, message);
       }
     }
   }
 }
 
+void ShenandoahVerifier::confirm_filled(HeapWord* start, HeapWord* end, const char *message) {
+  ShenandoahGenerationalHeap* gen_heap = ShenandoahGenerationalHeap::heap();
```
Comment (earthling-amzn) [inline src/hotspot/share/gc/shenandoah/shenandoahVerifier.cpp:1345]: `ShenandoahGenerationalHeap::heap` asserts that the mode is generational.

File: src/hotspot/share/gc/shenandoah/shenandoahVerifier.cpp [hunk 3]
```diff
@@ -1333,12 +1333,40 @@ void ShenandoahVerifier::help_verify_region_rem_set(ShenandoahHeapRegion* r, She
...
+  assert(gen_heap->mode()->is_generational(), "Filled is for generational");
+  ShenandoahOldGeneration* old_gen = gen_heap->old_generation();
+  ShenandoahOldHeuristics* old_heuristics = (ShenandoahOldHeuristics*) old_gen->heuristics();
```
Comment (earthling-amzn) [inline src/hotspot/share/gc/shenandoah/shenandoahVerifier.cpp:1348]: Cast shouldn't be necessary here.

File: src/hotspot/share/gc/shenandoah/shenandoahVerifier.cpp [hunk 4]
```diff
@@ -1333,12 +1333,40 @@ void ShenandoahVerifier::help_verify_region_rem_set(ShenandoahHeapRegion* r, She
...
+    oop obj = cast_to_oop(start);
+
+    // This object is not marked.  It should be a fill object, or an object that is about to be filled.  If it is a fill
+    // object, it should be array of int.
+
+    // if there are coalesce-and-fill candidates, skip over this object.  Otherwise, this regions has already been
+    // coalesced and filled, so this should be a fill object.
+    if (check_fill_objects && !obj->is_array()) {
+      log_info(gc)("confirm_filled thinks thinks is_old_bitmap_stable(): %s, old_gen->is_mark_complete(): %s",
+                   gen_heap->is_old_bitmap_stable()? "yes": "no", old_gen->is_mark_complete()? "yes": "no");
```
Comment (earthling-amzn) [inline src/hotspot/share/gc/shenandoah/shenandoahVerifier.cpp:1361]: Shoot, I just integrated https://github.com/openjdk/shenandoah/pull/415, it replaces `shHeap::is_old_bitmap_stable` with `shOldGen::is_mark_complete`.

File: src/hotspot/share/gc/shenandoah/shenandoahVerifier.cpp [hunk 5]
```diff
@@ -1333,12 +1333,40 @@ void ShenandoahVerifier::help_verify_region_rem_set(ShenandoahHeapRegion* r, She
...
+      ShenandoahAsserts::print_failure(ShenandoahAsserts::_safe_all, obj, start, nullptr, message,
```
Comment (earthling-amzn) [inline src/hotspot/share/gc/shenandoah/shenandoahVerifier.cpp:1362]: Why not `assert` here?

**Review by earthling-amzn — CHANGES_REQUESTED** [review]
Maybe something for a different PR, but can we now also stop swapping the remembered set for the verifier for global collections? We do this here: https://github.com/openjdk/shenandoah/blob/master/src/hotspot/share/gc/shenandoah/shenandoahConcurrentGC.cpp#L622

File: src/hotspot/share/gc/shenandoah/shenandoahVerifier.cpp (hunk 4 above)
Comment (kdnilsen) [reply src/hotspot/share/gc/shenandoah/shenandoahVerifier.cpp:1361]: merged in and resolved conflict.

File: src/hotspot/share/gc/shenandoah/shenandoahVerifier.cpp (hunk 3 above)
Comment (kdnilsen) [reply src/hotspot/share/gc/shenandoah/shenandoahVerifier.cpp:1348]: Thanks.  Removed.

File: src/hotspot/share/gc/shenandoah/shenandoahVerifier.cpp (hunk 2 above)
Comment (kdnilsen) [reply src/hotspot/share/gc/shenandoah/shenandoahVerifier.cpp:1345]: Thanks.  Removed the extra assert.

File: src/hotspot/share/gc/shenandoah/shenandoahVerifier.cpp (hunk 1 above)
Comment (kdnilsen) [reply src/hotspot/share/gc/shenandoah/shenandoahVerifier.cpp:1383]: Thanks.  Changing.

File: src/hotspot/share/gc/shenandoah/shenandoahVerifier.cpp (hunk 5 above)
Comment (kdnilsen) [reply src/hotspot/share/gc/shenandoah/shenandoahVerifier.cpp:1362]: Asserts:print_failure() makes it easier to provide more diagnostic information than would be provided by an assert failures.

**Pull request closed without merging**